File: src/codex_team/daemon/compaction.py
from __future__ import annotations

from codex_team.config import Config
from codex_team.daemon.event_bus import EventBus
from codex_team.daemon.registry import RegistryStore


class CompactionMonitor:
    def __init__(self, *, cfg: Config, registry: RegistryStore, event_bus: EventBus):
        self._cfg = cfg
        self._registry = registry
        self._event_bus = event_bus
        self._suggested: set[str] = set()

    async def observe_usage(self, name: str, tokens: int) -> None:
        if tokens < self._cfg.compaction.threshold_tokens:
            return
        if name in self._suggested:
            return
        self._suggested.add(name)
        self._event_bus.publish(
            "events",
            {
                "kind": "compact-suggest",
                "session": name,
                "tokens": tokens,
                "threshold": self._cfg.compaction.threshold_tokens,
            },
        )

    def clear(self, name: str) -> None:
        self._suggested.discard(name)

File: src/compaction.ts
import { Config } from "./config";
import { EventBus } from "./eventBus";
import { RegistryStore } from "./registry";

export class CompactionMonitor {
  private readonly suggestedLevel = new Map<string, number>();

  constructor(
    private readonly cfg: Config,
    private readonly registry: RegistryStore,
    private readonly eventBus: EventBus,
  ) {
    void this.registry;
  }

  async observeUsage(
    name: string,
    usage: {
      contextTokensEstimate: number | null;
      modelContextWindow: number | null;
      cumulativeUsageTokens: number | null;
    },
  ): Promise<void> {
    const metric = usage.contextTokensEstimate ?? usage.cumulativeUsageTokens ?? 0;
    const threshold = this.cfg.compaction.thresholdTokens;
    if (metric < threshold) {
      return;
    }
    const level = Math.max(1, Math.floor(metric / threshold));
    const previousLevel = this.suggestedLevel.get(name) || 0;
    if (level <= previousLevel) {
      return;
    }
    this.suggestedLevel.set(name, level);
    this.eventBus.publish("events", {
      kind: "compact-suggest",
      session: name,
      tokens: metric,
      level,
      metric_kind: usage.contextTokensEstimate != null ? "context_estimate" : "cumulative_usage",
      context_tokens_estimate: usage.contextTokensEstimate,
      model_context_window: usage.modelContextWindow,
      cumulative_usage_tokens: usage.cumulativeUsageTokens,
      threshold,
    });
  }

  clear(name: string): void {
    this.suggestedLevel.delete(name);
  }
}
